src/darwin/Framework/CHIP/CHIPDeviceController.mm

```diff
@@ -134,328 +141,132 @@ - (instancetype)init
-    chip::System::PacketBuffer::Free(buffer);
-}
-
-static void onInternalError(chip::DeviceController::ChipDeviceController * deviceController, void * appReqState, CHIP_ERROR error,
-    const chip::Inet::IPPacketInfo * pi)
-{
-    CHIPDeviceController * controller = (__bridge CHIPDeviceController *) appReqState;
-    [controller _dispatchAsyncErrorBlock:[CHIPError errorForCHIPErrorCode:error]];
-}
-
-- (void)_dispatchAsyncErrorBlock:(NSError *)error
-{
-    CHIP_LOG_METHOD_ENTRY();
-
-    id<CHIPDeviceControllerDelegate> strongDelegate = [self delegate];
-    if (strongDelegate && [self delegateQueue]) {
-        dispatch_async(self.delegateQueue, ^{
-            [strongDelegate deviceControllerOnError:error];
-        });
-    }
-}
-
-- (void)_dispatchAsyncMessageBlock:(NSData *)data
-{
-    CHIP_LOG_METHOD_ENTRY();
-
-    id<CHIPDeviceControllerDelegate> strongDelegate = [self delegate];
-    if (strongDelegate && [self delegateQueue]) {
-        dispatch_async(self.delegateQueue, ^{
-            [strongDelegate deviceControllerOnMessage:data];
-        });
+        _lastPairedDeviceID = 0;
     }
+    return self;
 }
 
-- (void)_dispatchAsyncConnectBlock
+// MARK: CHIPDeviceStatusDelegate
+- (void)onMessageReceived:(NSData *)message
 {
     CHIP_LOG_METHOD_ENTRY();
 
     id<CHIPDeviceControllerDelegate> strongDelegate = [self delegate];
     if (strongDelegate && [self delegateQueue]) {
         dispatch_async(self.delegateQueue, ^{
-            [strongDelegate deviceControllerOnConnected];
+            [strongDelegate deviceControllerOnMessage:message];
         });
     }
 }
 
-- (BOOL)connect:(NSString *)ipAddress error:(NSError * __autoreleasing *)error
-{
-    // Start the IO pump
-    dispatch_async(_chipSelectQueue, ^() {
-        self.cppController->ServiceEvents();
-    });
-    return YES;
-}
-
-- (BOOL)connect:(uint16_t)discriminator setupPINCode:(uint32_t)setupPINCode error:(NSError * __autoreleasing *)error
+- (BOOL)pairDevice:(uint64_t)deviceID
+     discriminator:(uint16_t)discriminator
+      setupPINCode:(uint32_t)setupPINCode
+             error:(NSError * __autoreleasing *)error
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
     [self.lock lock];
 
     chip::RendezvousParameters params = chip::RendezvousParameters().SetSetupPINCode(setupPINCode).SetDiscriminator(discriminator);
-    err = self.cppController->ConnectDevice(
-        kRemoteDeviceId, params, (__bridge void *) self, onConnected, onMessageReceived, onInternalError);
+    err = self.cppController->PairDevice(deviceID, params);
     [self.lock unlock];
 
     if (err != CHIP_NO_ERROR) {
-        CHIP_LOG_ERROR("Error(%d): %@, connect failed", err, [CHIPError errorForCHIPErrorCode:err]);
+        CHIP_LOG_ERROR("Error(%d): %@, failed in pairing the device", err, [CHIPError errorForCHIPErrorCode:err]);
         if (error) {
             *error = [CHIPError errorForCHIPErrorCode:err];
         }
         return NO;
     }
 
+    _lastPairedDeviceID = deviceID;
```

Comment:
We should only set this if pairing succeeds? i.e maybe track it as `currentlyPairingDeviceID` while pairing is in progress and then set `_lastPairedDeviceID = currentlyPairingDeviceID` and unset `currentlyPairingDeviceID` after pairing is done, stopped, or failed

Reply:
Yeah, once we update the rendezvous delegate callback to include the device ID, we can move it there.

Reply:
I'm suggesting we track it inside this file in the meantime. That way we don't have to wait for it to come back from Rendezvous?

Reply:
Actually, thought about it more. I'll remove this function altogether from the framework, and move this logic to the app.

Reply:
I see, I think most apps will need to track the pairing state, I might be more convenient to have the framework track that. wdyt?

Reply:
Sagar and I discussed it offline, and tracking it in framework requires framework to add itself as delegate. Currently only application is the delegate, and framework is a pass thru.
We will reconsider it in future if framework requires to be delegate for some other functionality.